src/rpmrebuild.1.in
.TH "rpmrebuild" "1" "@VERSION@" "Eric Gerbier" "rpm tools"
.SH "NAME"
rpmrebuild \- a tool to build an rpm package from an installed package
.SH "SYNOPSIS"
rpmrebuild [\fIoption\fP] <\fIpackagename\fP>
.br 
or
.br 
rpm \-\-rpmrebuild [\fIoption\fP] <\fIpackagename\fP>
.SH "DESCRIPTION"
rpmrebuild is a tool to build easily rpm package.
it can be used to build an rpm file from an installed package (lost rpm) or 
to quickly make change to a package: 
just have your change on installed files and call rpmrebuild.
.SH "OPTIONS"
.TP 
\fB\-a, \-\-additional\fP=<\fIflags\fP>
pass additional flags to \fBrpmbuild\fP
.TP 
\fB\-b, \-\-batch\fP
work in batch mode (no interactive query if files change)
.TP 
\fB\-c, \-\-comment\-missing\fP=<\fIyes\fP|\fIno\fP>
comment out in the specfile missing files. Default: \fBno\fP.
.TP 
\fB\-D, \-\-define\fP=<\fIdefines\fP>
defines to be passed to rpmbuild.
.TP 
\fB\-d, \-\-directory\fP=<\fIdirectory\fP>
rpm package result will be written in \fIdirectory\fP.
.TP 
\fB\-e, \-\-edit\-spec\fP
alias to \-\-edit\-whole.
.TP 
\fB\-f, \-\-filter\fP=<\fIcommand\fP>
alias for \fB\-\-change\-spec\-whole\fP
.TP 
\fB\-h, \-\-help\fP
output help information about all, but plugins option and exit.
.TP 
\fB\-\-help\-plugins\fP
output help information about plugins options and exit.
.TP 
\fB\-i, \-\-include\fP=<\fIfile\fP>
read more options from the file 'file'.
.TP 
\fB\-k, \-\-keep\-perm
alias to \fB\-\-pug\-from\-fs\fP
.TP 
\fB\-m, \-\-modify\fP=<\fIcommand\fP>
alias to \fB\-\-change\-files
.TP 
\fB\-n, \-\-notest\-install \fP
do not test install at end of rpm building
.TP 
\fB\-p, \-\-package\fP
use package file, not installed rpm (this option is like rpm's \-p option)
.TP 
\fB\-\-pug\-from\-db (default)\fP
set files permission, uid and gid as saved in rpm database
.TP 
\fB\-\-pug\-from\-fs\fP
keep installed files permission, uid and gid 
(do not force to original files permission, uid and gid saved in rpm database)
.TP 
\fB\-P, \-\-autoprovide\fP
use rpm scripts to compute provides
.TP 
\fB\-R, \-\-autorequire\fP
use rpm scripts to compute requires
.TP 
\fB\-r, \-\-release\fP=<\fIversion\fP>\fP
specify a release
.TP 
\fB\-s, \-\-spec\-only\fP=<\fIspecfile\fP>
do not rebuild package, only generate specfile. 
Can't be used with \fB\-\-change\-files\fP.
.TP 
\fB\-v, \-\-verbose\fP
verbose output
.TP 
\fB\-\-debug\fP
for debugging purposes : do not remove temporary files at end of work
.TP 
\fB\-V, \-\-version\fP
output version information and exit.
.TP 
\fB\-w, \-\-warning\fP
display some warnings if it detects a potential problem 
(for example, spec file which contains filename with globbing char).
.TP 
\fB\-y, \-\-verify\fP=<\fIyes\fP|\fIno\fP>
Turn on/off package verifying before processing. 
Default: \fByes\fP. Can be turn off to get better performance.
.LP 
PLUGINS OPTION
.LP 
rpmrebuild use may be extended with plugins.
Plugins are searched in \fB/usr/lib/rpmrebuild/plugins\fP and then in $PATH
.TP 
\fB\-\-change\-files\fP=<\fIcommand\fP>
Invoke plugin which change package's files. Note, that these files are copied
before to the temporary location, so installed package's files not changed. 
.br 
Can't be used with \fB\-\-spec\-only\fP
.LP 
\fB\-\-change\-spec\-whole\fP=<\fIcommand\fP>
.br 
\fB\-\-edit\-whole
.RS
Invoke plugin or editor to change specfile.
.RE
.LP 
\fB\-\-change\-spec\-preamble\fP=<\fIcommand\fP>
.br 
\fB\-\-edit\-preamble
.LP 
\fB\-\-change\-spec\-conflicts\fP=<\fIcommand\fP>
.br 
\fB\-\-edit\-conflicts
.LP 
\fB\-\-change\-spec\-obsoletes\fP=<\fIcommand\fP>
.br 
\fB\-\-edit\-obsoletes
.LP 
\fB\-\-change\-spec\-provides\fP=<\fIcommand\fP>
.br 
\fB\-\-edit\-provides
.LP 
\fB\-\-change\-spec\-requires\fP=<\fIcommand\fP>
.br 
\fB\-\-edit\-requires
.LP 
\fB\-\-change\-spec\-description\fP=<\fIcommand\fP>
.br 
\fB\-\-edit\-description
.LP 
\fB\-\-change\-spec\-files\fP=<\fIcommand\fP>
.br 
\fB\-\-edit\-files
.LP 
\fB\-\-change\-spec\-triggers\fP=<\fIcommand\fP>
.br 
\fB\-\-edit\-triggers
.LP 
\fB\-\-change\-spec\-pre\fP=<\fIcommand\fP>
.br 
\fB\-\-edit\-pre
.LP 
\fB\-\-change\-spec\-post\fP=<\fIcommand\fP>
.br 
\fB\-\-edit\-post
.LP 
\fB\-\-change\-spec\-preun\fP=<\fIcommand\fP>
.br 
\fB\-\-edit\-preun
.LP 
\fB\-\-change\-spec\-postun\fP=<\fIcommand\fP>
.br 
\fB\-\-edit\-postun
.LP 
\fB\-\-change\-spec\-verifyscript\fP=<\fIcommand\fP>
.br 
\fB\-\-edit\-verifyscript
.LP 
\fB\-\-change\-spec\-changelog\fP=<\fIcommand\fP>
.br 
\fB\-\-edit\-changelog
.LP 
.RS
Invoke plugin or editor to change preamble, conflicts, obsoletes, provides,
requires, description, files, triggers, preinstall, postinstall, preuninstall,
postuninstall, verifyscript or changelog.
.br 
None of these options can be specified after \fB\-\-change\-spec\-whole\fP
or \fB\-\-edit\-whole\fP.
.SH "INCLUDE FILES"
Any of the rpmrebuild's LONG option can be specified on the command line or
in the include file (file specified with \-i/\-\-include option).
.LP 
Include files are also searched in the \fB/usr/lib/rpmrebuild/plugins\fP directory.
.LP 
Include file has following syntax.
.TP 
\fBcomments\fP
Any line begin with '#' or some number of whitespace and '#'.
.TP 
\fBempty line\fP
empty line are ignored.
.TP 
\fBoption line\fP
option line is very similar to the option written on the command line.
the limitation of the option line in the file are:
.RS
.TP 
long option
should be used ONLY long options WITHOUT leading \-\-.
.TP 
option's argument
option's argument should not be quoted. 
.TP 
separator
option's argument should be separated from option by spaces or tabs, not by
equal ('=') sign.
.TP 
non\-optional argument(s)
package name (non\-optional argument) shouldn't be specified in the include file.
.RE
.LP 
For example, if command line looks like:
.RS
rpmrebuild \-\-change\-files="pgm1 a b c" \-\-change\-spec\-pre="pgm2 bb" pkg
.RE
.LP 
Then include file should be written like:
.RS
change\-files    pgm1 a b c
.br 
change\-spec\-pre pgm2 bb
.br 
.RE
.SH "FILES"
.TP 
\fB~/.rpmmacros\fP 
is used by rpm when building packages (macros)
.TP 
\fB/etc/popt\fP
contains rpm alias used by rpmrebuild
.TP 
\fB$RPMREBUILD_TMPDIR/work\fP
directory where rpmrebuild put temporary files.
.SH "ENVIRONMENT"
.LP 
.TP 
\fBRPMREBUILD_TMPDIR\fP
Directory for temporal files. If not set ~/.tmp/rpmrebuild.$$ is used.
.TP 
\fBVISUAL\fP
used to specify an editor for specfile. if not set \fBEDITOR\fP variable examined.
.TP 
\fBEDITOR\fP
used to specify an editor for specfile. if not set \fBvi\fP used.
.SH "EXAMPLES"
.LP 
Build an rpm package as closed as possible to original one:
.LP 
\fBrpm \-\-rpmrebuild \-\-batch package\fP
.LP 
Build a modified package:
.LP 
\fBrpm \-\-rpmrebuild \-\-keep\-perm package\fP
.LP 
Build a rpm from another rpm file:
.LP 
\fBrpm \-\-rpmrebuild \-\-package \-\-filter filter /usr/src/redhat/i386/package\-i386.rpm\fP
.LP 
Rebuild rpm as legacy (rpm[23]) package:
.LP 
\fBrpm \-\-rpmrebuild \-\-additional "\-\-nodirtokens" bash\fP
.LP 
Specify generated rpm name:
.LP 
\fBrpm \-\-rpmrebuild \-\-define "_rpmfilename %%{NAME}.rpm" bash\fP
.LP 
Change package interactive
.LP 
\fBrpm \-\-rpmrebuild \-\-package \-\-change\-files "/bin/bash" /some/place/bash.rpm\fP
.LP 
Change package non\-interactive
.LP 
\fBrpm \-\-rpmrebuild \-\-package \-\-change\-files "a.sh && b.sh" /some/place/bash.rpm\fP
.LP 
if you have problem with rpm building, and if you know specfile syntax, you can also edit it:
.LP 
\fBrpm \-\-rpmrebuild  \-\-edit\-whole package\fP
.SH "AUTHORS"
.LP 
Eric Gerbier <gerbier@users.sourceforge.net>
.br 
Valery Reznic <valery_reznic@users.sourceforge.net>
.TP 
See <URL:http://rpmrebuild.sourceforge.net/>.
.SH "NOTES"
.LP 
.TP 
rpmrebuild can not reproduce exactly a package, so
.br 
\- md5 checksum will not match original ones.
.br 
\- gpg or pgp signatures are lost

.TP 
packages which contains file with globbing characters (* .. ? ) will not build
.SH "COPYRIGHT"
.LP 
rpmrebuild is copyright (C) 2002 by Eric Gerbier.
.br 
this program is distributed under GNU General Public License
.br 
It is provided "as is", without any express or implied
warranties.  
.br
See the file COPYING for details.
.SH "SEE ALSO"
.LP 
rpm(8), rpmbuild(8), rpmrebuild_plugins(1), 'man \-k rrp' for plugins specific manpages.



src/News
* 2005-03-04 : release 2.0.2-1, bug fix for 2.x releases (BuildArch)
* 2005-02-25 : release 2.0.1-1, fix for rpm-4.3
* 2005-02-21 : release 2.0.0-1, major code rewrite to be more modular, allow change in specific spec sections
* 2004-08-16 : release 1.4.7-3, dependency changes to work on suse, rpmlint check, bug fix
* 2004-07-20 : release 1.4.7-1, bugs fix, working on file packages, avoid some sed treatments
* 2004-05-24 : release 1.4.6-4, security: change working directory to ~/.tmp/rpmrebuild, bug fix
* 2004-05-19 : release 1.4.6-3, add remaining dependencies, long option abbreviation
* 2004-05-06 : release 1.4.6-2, add rpm-build dependency for rpm > 4.0
* 2004-01-01 : release 1.4.6-1, add --autoprovide and --autorequire flags
* 2003-11-14 : release 1.4.5-2, bug fix
* 2003-07-10 : release 1.4.5-1, add some flags to improve performances (-y/-c)
* 2003-06-18 : release 1.4.4-1, now can modify an rpm file
* 2003-06-10 : release 1.4.3-1, to work on rpm 4.2 (redhat 9 for exemple)


src/rpmrebuild.sh
#!/bin/bash
###############################################################################
#   rpmrebuild.sh 
#
#    Copyright (C) 2002 by Eric Gerbier
#    Bug reports to: gerbier@users.sourceforge.net
#    $Id$
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
###############################################################################
VERSION="$Id$"

# debug 
#set -x 
###############################################################################
function Echo
{
	echo -e "$@" 1>&2
}
###############################################################################
function Error
{
	Echo "$0: ERROR: $@"
}
###############################################################################

function Warning
{
	# screen
	Echo "$0: WARNING: $@"
	# for optionnal bug report
	echo -e "$0: WARNING: $@" >> $BUGREPORT

}
###############################################################################

function AskYesNo
{
	local Ans
	echo -en "$@ ? (y/N) " 1>&2
	read Ans
	case "x$Ans" in
		x[yY]*) return 0;;
		*)      return 1;;
	esac || return 1 # should not happened
	return 1 # should not happend
}

###############################################################################
function RmDir
{
	[ $# -ne 1 -o "x$1" = "x" ] && {
		Echo "Usage: RmDir <dir>"
		return 1
	}
	# to ensure tmpdir is really emptied by rm -rf
	local Dir
	Dir="$1"
	if [ -d $Dir ]
	then
		rm -rf "$Dir" 2>/dev/null && return
		chmod -R 700 "$Dir" 2>/dev/null  # no return here !!!
		rm -rf "$Dir" || return
	fi
	return 0
}
###############################################################################
function SpecEdit
{
	[ $# -ne 1 -o "x$1" = "x" ] && {
		Echo "Usage: $0 SpecEdit <file>"
		return 1
	}
	# -e option : edit the spec file
	local File=$1
	${VISUAL:-${EDITOR:-vi}} $File
	AskYesNo "$WantContinue" || {
		Aborted="yes"
		Echo "Aborted."
	        return 1
	}
	return 0
}
###############################################################################

function VerifyPackage
{
	# verification des changements
	# check for package change
	rpm --verify --nodeps ${PAQUET} # Don't return here, st=1 - verify fail 
	return 0
}

###############################################################################
function QuestionsToUser
{
	[ "X$batch"     = "Xyes" ] && return 0 ## batch mode, continue
	[ "X$spec_only" = "Xyes" ] && return 0 ## spec only mode, no questions

	AskYesNo "$WantContinue" || return
	RELEASE_ORIG="$(spec_query_qf '%{RELEASE}')"
	[ -z "$RELEASE_NEW" ] && \
	AskYesNo "$WantChangeRelease" && {
		echo -n "$EnterRelease $RELEASE_ORIG): "
		read RELEASE_NEW
	}
	return 0
}

###############################################################################
function IsPackageInstalled
{
	# test if package exists
	local output
	output="$( rpm --query ${PAQUET} 2>&1 )" # Don't return here - use output
	iret=$?
	if [ $iret -eq 1 ]
	then
		# no such package in rpm database
		Error "${PAQUET} $PackageNotInstalled"
		return 1
	else
		# find it : one or more ?
		set -- $output
		case $# in
			1)
			: # Ok, do nothing
			;;

			*)
			Error "$PackageTooMuch '${PAQUET}':\n$output"
			return 1
			;;
		esac 
	fi
	return 0
}

###############################################################################
function RpmUnpack
{
	[ "x$BUILDROOT" = "x/" ] && {
		Error "$BuildRootError" 
        	return 1
	}
	local CPIO_TEMP=$TMPDIR_WORK/${PAQUET_NAME}.cpio
	rm --force $CPIO_TEMP                               || return
	rpm2cpio ${PAQUET} > $CPIO_TEMP                     || return
	rm    --force --recursive $BUILDROOT                || return
	mkdir --parent            $BUILDROOT                || return
	(cd $BUILDROOT && cpio --quiet -idmu ) < $CPIO_TEMP || return
	rm --force $CPIO_TEMP                               || return
	# Process ghost files
	/bin/bash $MY_LIB_DIR/rpmrebuild_ghost.sh $BUILDROOT < $FILES_IN || return
	return 0
}
###############################################################################
function CreateBuildRoot
{
        if [ "x$package_flag" = "x" ]; then
		if [ "X$need_change_files" = "Xyes" ]; then
			/bin/bash $MY_LIB_DIR/rpmrebuild_buildroot.sh $BUILDROOT < $FILES_IN || return
		else
			: # Do nothing
		fi
	else
        	RpmUnpack || return
	fi 
	return 0
}
###############################################################################

function RpmBuild
{
	# reconstruction fichier rpm : le src.rpm est inutile
	# build rpm file, the src.rpm is not usefull to do
	# for rpm 4.1 : use rpmbuild
	local BUILDCMD=rpm
	[ -x /usr/bin/rpmbuild ] && BUILDCMD=rpmbuild
	eval $BUILDCMD $rpm_defines -bb $rpm_verbose $additional ${FIC_SPEC} || {
   		Error "package '${PAQUET}' $BuildFailed"
   		return 1
	}
	
	return 0
}

###############################################################################
function RpmFileName
{
	QF_RPMFILENAME=$(eval rpm $rpm_defines --eval %_rpmfilename) || return
	RPMFILENAME=$(eval rpm $rpm_defines --specfile --query --queryformat "${QF_RPMFILENAME}" ${FIC_SPEC}) || return
	# workarount for redhat 6.x
	arch=$(eval rpm $rpm_defines --specfile --query --queryformat "%{ARCH}"  ${FIC_SPEC})
	if [ $arch = "(none)" ]
	then
		arch=$(eval rpm $rpm_defines --query $package_flag --queryformat "%{ARCH}" ${PAQUET})
		RPMFILENAME=$(echo $RPMFILENAME | sed "s/(none)/$arch/g")
	fi

	[ -n "$RPMFILENAME" ] || return
	RPMFILENAME="${rpmdir}/${RPMFILENAME}"
	return 0
}

###############################################################################
function InstallationTest
{
	# installation test
	# force is necessary to avoid the message : already installed
	rpm -U --test --force ${RPMFILENAME} || {
		Error "package '${PAQUET}' $TestFailed"
		return 1
	}
	return 0
}

###############################################################################
function Processing
{
	local Aborted="no"
	local MsgFail

	source $RPMREBUILD_PROCESSING && return 0

	if [ "X$need_change_spec" = "Xyes" -o "X$need_change_files" = "Xyes" ]; then
		[ "X$Aborted" = "Xyes" ] || Error "package '$PAQUET' $ModificationFailed."
	else
		Error "package '$PAQUET' $SpecFailed."
	fi
	return 1
}
###############################################################################
# recover system informations on rpmrebuild context
function GetInformations
{
	Echo "from: $1"
	Echo "-----------"
	lsb_release -a
	cat /etc/issue
	Echo "-----------"
	rpm -q rpmrebuild
	rpm -q rpm
	Echo "-----------"
	rpm --querytags
	Echo " --------------- $WriteComments -----------------------"
}
###############################################################################
# send informations to developper to allow fix problems
function SendBugReport
{
	[ "X$batch"     = "Xyes" ] && return 0 ## batch mode, skip report
	AskYesNo "$WantSendBugReport" || return
	# build default mail address 
	from="${USER}@${HOSTNAME}"
	AskYesNo "$WantChangeEmail ($from)" && {
		echo -n "$EnterEmail"
		read from
	}
	GetInformations $from >> $BUGREPORT 2>&1
	AskYesNo "$WantEditReport" && {
		${VISUAL:-${EDITOR:-vi}} $BUGREPORT
	}
	AskYesNo "$WantStillSend" && {
		mail -s "[rpmrebuild] bug report" rpmrebuild-bugreport@lists.sourceforge.net < $BUGREPORT
	}
	return
}
###############################################################################
# rpm tags change along the time : some are added, some are renamed, some are
# deprecated, then removed
# the idea is to check if the tag we use for rpmrebuild still exists
function CheckTags
{
	# get rpm tags
	rpm_tags=$( rpm --querytags )

	# get tags used in rpmrebuild
	# rpmrebuild_popt should follow a strict syntaxe
	# "%|" begin a test line
	# tr command remove all not alpha characters
	# the "dummy" query should be filtered too
	rpmrebuild_tags=$( grep "^%|" $MY_LIB_DIR/rpmrebuild_popt |  tr -cs '[:alpha:]' '[\n*]' | grep -v dummy )

	# check for all rpmrebuild tags
	errors=0
	for tag in $rpmrebuild_tags
	do
		ok=''
		# if it exists in rpm tags
		for rpm_tag in $rpm_tags
		do
			if [ "$tag" = "$rpm_tag" ]
			then
				# ok : we find it
				ok='y'
				break
			fi	
		done
		if [ -z "$ok" ]
		then
			Warning "(CheckTags) $MissingTag $tag"
			let errors="$errors + 1"
		fi
	done
	if [ $errors -ge 1 ] 
	then
		Warning "$CannotWork"
		SendBugReport
		return 1
	fi
}
##############################################################
# Main Part                                                  #
##############################################################
# shell pour refabriquer un fichier rpm a partir de la base rpm
# a shell to build an rpm file from the rpm database

function Main
{
	
	#WantContinue="Do you want to continue"

	RPMREBUILD_TMPDIR=${RPMREBUILD_TMPDIR:-~/.tmp/rpmrebuild.$$}
	#RPMREBUILD_TMPDIR=${RPMREBUILD_TMPDIR:-~/.tmp/rpmrebuild}
	BUGREPORT=$RPMREBUILD_TMPDIR/bugreport
	export BUGREPORT
	export RPMREBUILD_TMPDIR
	TMPDIR_WORK=$RPMREBUILD_TMPDIR/work

	# create tempory directories before any work/test
	RmDir "$RPMREBUILD_TMPDIR" || return
	mkdir -p $TMPDIR_WORK       || return

	FIC_SPEC=$TMPDIR_WORK/spec
	FILES_IN=$TMPDIR_WORK/files.in
	# I need it here just in case user specify 
	# plugins for fs modification  (--change-files)
	BUILDROOT=$TMPDIR_WORK/root

	MY_LIB_DIR=`dirname $0` || return
	source $MY_LIB_DIR/rpmrebuild_parser.src || return
	source $MY_LIB_DIR/spec_func.src         || return
	source $MY_LIB_DIR/processing_func.src   || return

	# check language
	lang=$( echo $LANG | cut -f 1 -d_)
	case "$lang" in
		en) real_lang=en;;
		fr) real_lang=fr;;
		*)  real_lang=en;;
	esac
	# load translation file
	source $MY_LIB_DIR/locale/$real_lang/rpmrebuild.lang
	
	processing_init || return
	CheckTags || return

	export RPMREBUILD_PLUGINS_DIR=${MY_LIB_DIR}/plugins

	# suite a des probleme de dates incorrectes
	# to solve problems of bad date
	export LC_TIME=POSIX

	RPMREBUILD_PROCESSING=$TMPDIR_WORK/PROCESSING

	CommandLineParsing "$@" || return
	[ "x$NEED_EXIT" = "x" ] || return $NEED_EXIT

	if [ "x" = "x$package_flag" ]; then
   		[ "X$need_change_files" = "Xyes" ] || BUILDROOT="/"
   		IsPackageInstalled || return
   		if [ "X$verify" = "Xyes" ]; then
      			out=$(VerifyPackage) || return
      			if [ -n "$out" ]; then
		 		Warning "$FilesModified\n$out"
		 		QuestionsToUser || return
      			fi
   		else # NoVerify
			:
   		fi
	else
		:
		# When rebuilding package from .rpm file it's just native
		# to use perm/owner/group from the package.
		# But because it anyway default and if one has a reason
		# to change it, one can. I am not force it here anymore.
		#RPMREBUILD_PUG_FROM_FS="no"  # Be sure use perm, owner, group from the pkg query.
	fi

	if [ "X$spec_only" = "Xyes" ]; then
		BUILDROOT="/"
		SpecGeneration   || return
		Processing       || return
	else
		SpecGeneration   || return
		CreateBuildRoot  || return
		Processing       || return
		RpmBuild         || return
		RpmFileName      || return
		echo "result: ${RPMFILENAME}"
		if [ -z "$NOTESTINSTALL" ]; then
			InstallationTest || return
		fi
	fi
	return 0
}
###############################################################################

Main "$@"
st=$?	# save status

# in debug mode , we do not clean temp files
if [ -z "$debug" ]
then
	RmDir "$RPMREBUILD_TMPDIR"
fi
exit $st

#####################################
# BUILDROOT note.
# My original idea was for recreating package from another rpm file
# (not installed) use 'rpm -bb --define "buildroot foo"', but
# It does not work:
#  when i not specify buildroot in the spec file default value is "/"
#  I can build this package, but can't override buildroot from the
#  command line.
#
# when i specify buildroot: / in the spec file i got parser error.
#
# So, for recreating installed packages I need specfile WITHOUT
# buildroot
# For recreating package from another rpm I have to put buildroot in the
# specfile
#########################################
